### DISCUSSION TITLE
Building GPUI without Xcode's `metal` and `metallib`

### DISCUSSION THREAD

**Discussion by niklaskorz** [zed-industries/zed#7016] [Feature Requests]
### Problem statement

Building Zed, particularly GPUI, in isolated environments such as Nix currently fails because the tools `metal` and `metallib` are only included inside `Xcode.app`, not in any macOS Frameworks or the command line tools. This has already been reported, see #6442.

### Solution proposal

Precompiling Metal shaders can be turned into a feature flag on the gpui crate. This feature flag can be turned on by default to not cause any issues with current behavior.

The [call to `compile_metal_shaders`](https://github.com/zed-industries/zed/blob/1313402a6bfd3f508aec8b4a8d89b0b5a826261b/crates/gpui/build.rs#L12) inside gpui's `build.rs` would be put behind a feature guard, as well as the [call to `device.new_library_with_data`](https://github.com/zed-industries/zed/blob/1313402a6bfd3f508aec8b4a8d89b0b5a826261b/crates/gpui/src/platform/mac/metal_renderer.rs#L65).

If the feature flag is not enabled, loading the shader should fall back to [`device.new_library_with_source`](https://developer.apple.com/documentation/metal/mtldevice/1433431-newlibrarywithsource).

### Pair programming

I'd be happy to join a pair programming session for implementing this! My time zone is UTC+1 (Germany).

**Comment by osiewicz** [comment]
Hey, that sounds like a good idea. I would be happy to pair on that. :)
I'm in UTC+1 as well. Do you have a specific time-slot in mind?

**Reply by niklaskorz** [reply]
Today (Tuesday) or tomorrow (Wednesday) between 15:00 and 20:00 would work for me.

**Reply by osiewicz** [reply]
Wednesday 15:00 would be sweet. :) Would you mind if I shared that date in Discord in case anybody else wants to join us as guests?

**Reply by niklaskorz** [reply]
Yup that's alright

**Comment by osiewicz** [comment]
Let's meet up in https://zed.dev/channel/gpui-365 in 25 minutes

**Reply by evanrichter** [reply]
Was there any progress made during the meeting?

**Reply by osiewicz** [reply]
Hi,
Yes, we ended up submitting a PR: https://github.com/zed-industries/zed/pull/7148.
Give that a try (`cargo build --features runtime shaders`) and let us know how it goes :)

**Comment by niklaskorz** [comment]
@osiewicz I have been digging some more and apparently it's possible to save precompiled archives at runtime, but those are a combination of pipelines + shaders, and are also specific to the particular GPU (e.g., can't share this binary between an M1 and an M2) and macOS version. This might be worth pursuing as it would also mean a potential improvement in startup performance even with already precompiled shaders. These binary archives are the _GPU-native_ binaries that end up being executed on the GPU, while the current "precompiling" (I'll call it offline compiling from now on) of the shaders just results in virtual machine bitcode that has to be further processed during runtime. If we had this, the initial startup time with runtime compiling would still be slightly worse than with offline compiling, but for subsequent starts (if we cache these native shader binaries somewhere), there should be no more difference between the runtime and offline compiled variants.

Here is the WWDC session regarding Metal binary archives: https://developer.apple.com/videos/play/wwdc2020/10615/

**Reply by osiewicz** [reply]
That's great news, would you like to pair on it sometime this week?

I'm not sure about this one bit:
> This might be worth pursuing as it would also mean a potential improvement in startup performance even with already precompiled shaders.

As in, I don't think pipeline building is particularly costly at runtime and caching the pipeline state would bring quite a bit of complexity in an of itself; it's fine to do for `runtime-shaders` feature though IMHO, as it seems to be mostly intended for use by developers of applications written with GPUI.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
